vm/src/warn.rs

```diff
@@ -72,32 +72,39 @@ fn setup_context(
     let __warningregistry__ = "__warningregistry__";
     let __name__ = "__name__";
 
-    // Setup globals, filename and lineno.
-    let frame = vm.current_frame();
+    let current_frame = vm.current_frame();
+    let mut f = current_frame.as_deref().cloned();
 
-    //     PyThreadState *tstate = _PyThreadState_GET();
-    //     PyFrameObject *f = PyThreadState_GetFrame(tstate);
-    //     // Stack level comparisons to Python code is off by one as there is no
-    //     // warnings-related stack level to avoid.
-    //     if (stack_level <= 0 || is_internal_frame(f)) {
-    //         while (--stack_level > 0 && f != NULL) {
-    //             PyFrameObject *back = PyFrame_GetBack(f);
-    //             Py_DECREF(f);
-    //             f = back;
-    //         }
-    //     }
-    //     else {
-    //         while (--stack_level > 0 && f != NULL) {
-    //             f = next_external_frame(f);
-    //         }
-    //     }
+    // Stack level comparisons to Python code is off by one as there is no
+    // warnings-related stack level to avoid.
 
-    let (globals, filename, lineno) = if let Some(f) = frame {
-        // TODO:
-        let lineno = 1;
-        // *lineno = PyFrame_GetLineNumber(f);
-        // *filename = code->co_filename;
-        (f.globals.clone(), f.code.source_path, lineno)
+    if let Some(frame) = current_frame {
+        if stack_level <= 0 || frame.is_internal_frame() {
+            while let Some(tmp) = frame.clone().f_back(vm) {
+                stack_level -= 1;
+                if stack_level > 0 {
+                    break;
+                }
+                f = Some(tmp);
+            }
+        } else {
+            stack_level -= 1;
+            if stack_level <= 0 {
+                f = frame.next_external_frame(vm);
+
+                while let Some(frame) = f.clone() {
+                    stack_level -= 1;
+                    if stack_level > 0 {
+                        break;
+                    }
+                    f = frame.next_external_frame(vm);
+                }
+            }
+        }
+    }
```

Comment:
```suggestion
    if stack_level <= 0 || f.clone().map(|frame| frame.is_internal_frame()).unwrap_or_default() {
        loop {
            stack_level -= 1;
            if stack_level <= 0 || f.is_none() {
                break;
            }
            f = f.unwrap().f_back(vm);
        }
    } else {
        loop {
            stack_level -= 1;
            if stack_level <= 0 || f.is_none() {
                break;
            }
            f = f.unwrap().next_external_frame(vm);
        }
    }
```

Reply:
I don't support `f.is_none()` and `f.unwrap()` combination when it doesn't give a significant advantage. It looks ok for now, but that works only when human carefully keep the pair in the correct order. We can write the same logic with `if let` matches which prevent mistakes from future logic changes.